src/main/java/fi/aalto/cs/intellij/common/BuildInfo.java

```diff
@@ -0,0 +1,123 @@
+package fi.aalto.cs.intellij.common;
+
+import fi.aalto.cs.intellij.util.StringSplitter;
+import java.util.Properties;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Holds information that was gathered during the build of the program (such as version).
+ * Normally, this class should be accessed via {@code BuildInfo.SingletonUtility.INSTANCE}.
+ */
+public class BuildInfo {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(BuildInfo.class);
+  private static final String RESOURCE_NAME = "build-info.properties";
+
+  public final Version version;
+
+  BuildInfo(@Nullable Properties props) {
+    Version ver = Version.fromProperties(props);
+    if (ver == null) {
+      ver = new Version(0,0,0);
+      LOGGER.error("Version information is invalid.  '{}' is used instead.", ver);
+    }
+    this.version = ver;
+  }
+
+  BuildInfo(@NotNull Resources resources) {
+    this(resources.getProperties(RESOURCE_NAME));
+  }
+
+  public static class Version {
+    private static final String PROPERTY_KEY = "version";
+
+    public final int major;
+    public final int minor;
+    public final int build;
+
+    /**
+     * Returns a {@link Version} object based on the value of {@value PROPERTY_KEY} in the given
+     * properties.
+     * @param props A {@link Properties} object, or {@code null} in which case {@code null} is also
+     *              returned.
+     * @return A {@link Version} object, or {@code null} if the version information could not be
+     *         read.
+     */
+    @Nullable
+    public static Version fromProperties(@Nullable Properties props) {
+      if (props == null) {
+        return null;
+      }
+      String version = props.getProperty(PROPERTY_KEY);
+      return Version.fromString(version);
+    }
+
+    /**
+     * Returns a {@link Version} object based on the value of the given string.
+     * @param version A version string of format "{major}.{minor}.{build}".
+     * @return A {@link Version} object, or {@code null} if {@code version} is invalid or null.
+     */
+    @Nullable
+    public static Version fromString(@Nullable String version) {
```

Comment:
JB themselves use [a simple String for version](https://github.com/JetBrains/intellij-community/blob/863153aec517a3c9b65bfd1993384f4304e79f78/platform/core-impl/src/com/intellij/ide/plugins/IdeaPluginDescriptorImpl.java#L75), but this is fancy!

Reply:
I implemented this kind of version class because we have to compare versions. In future, when we have to check if this course XML is compatible with this version of plugin, we can perhaps use some kind of semantic versioning: e.g. if course requires plugin version "1.6", all versions "1.6", "1.7", "1.8" are fine but "2.0" is not (because a new major version implies breaking changes). But this is something we don't have to care about now.

Reply:
hm... @OlliKiljunen check [this](https://github.com/JetBrains/intellij-community/blob/e6de8d8c64e5e2fbacf43c9abd7bc2475427239e/platform/platform-impl/src/com/intellij/openapi/updateSettings/impl/PluginDownloader.java#L215-L217) out

Reply:
A nice discovery. :+1

I am, however, skeptical if we should rely on that. It is seems to be poorly documented and use a lot of "smart" logic to handle many different version string patterns, which actually is more a downside for us because we want to our version strings to match only a certain pattern decided by us. What is more, there seems to be no easy way to check e.g. "is the same major version", which is something that we might need (because major versions usually imply breaking changes so for example a plugin v2.0 does not necessarily support files created for v1.18 so perhaps a warning should be logged/shown in that situation).

Link to the implementation:
https://github.com/JetBrains/intellij-community/blob/master/platform/util-rt/src/com/intellij/util/text/VersionComparatorUtil.java

There they say:
> See TeamCity documentation on requirements comparison for formal description.

Only documentation I was able to find quickly:
https://www.jetbrains.com/help/teamcity/configuring-agent-requirements.html

There is also a "dumber" version comparer:
https://github.com/JetBrains/intellij-community/blob/92e4c7785bc46fcdda7e56ee0d808df287be98c7/platform/util/src/com/intellij/openapi/util/text/StringUtil.java#L2643

Based on their comments on that file,  it seems to be something that can be deprecated in the future.